rasa/utils/tensorflow/transformer.py

```diff
@@ -0,0 +1,478 @@
+from typing import List, Optional, Text, Tuple, Callable
+import tensorflow as tf
+import tensorflow_addons as tfa
+from tensorflow.python.keras.utils import tf_utils
+import numpy as np
+from rasa.utils.tensorflow.layers import DenseWithSparseWeights
+
+
+# from https://www.tensorflow.org/tutorials/text/transformer
+# and https://github.com/tensorflow/tensor2tensor
+class MultiHeadAttention(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        d_model: int,
+        num_heads: int,
+        attention_dropout_rate: float = 0.0,
+        unidirectional: bool = False,
+        use_key_relative_position: bool = False,
+        use_value_relative_position: bool = False,
+        max_relative_position: Optional[int] = None,
+        heads_share_relative_embedding: bool = False,
+    ) -> None:
+        super().__init__()
+
+        self.num_heads = num_heads
+        self.d_model = d_model
+        self.attention_dropout_rate = attention_dropout_rate
+        self.unidirectional = unidirectional
+        self.use_key_relative_position = use_key_relative_position
+        self.use_value_relative_position = use_value_relative_position
+        self.relative_length = max_relative_position
+        if self.relative_length is not None:
+            self.relative_length += 1  # include current time
+        self.heads_share_relative_embedding = heads_share_relative_embedding
+
+        assert d_model % self.num_heads == 0
+
+        self._depth = d_model // self.num_heads
+
+        self._wq = DenseWithSparseWeights(units=d_model, use_bias=False)
+        self._wk = DenseWithSparseWeights(units=d_model, use_bias=False)
+        self._wv = DenseWithSparseWeights(units=d_model, use_bias=False)
+
+        self._dense = DenseWithSparseWeights(units=d_model)
+
+        self._create_relative_embeddings()
+
+    def _create_relative_embeddings(self) -> None:
+        """Create relative embeddings."""
+
+        if self.use_key_relative_position or self.use_value_relative_position:
```

Comment:
Minor:
We could get rid of the `else` clauses by first doing
```
relative_embedding_shape = None
self.key_relative_embeddings = None
self.value_relative_embeddings = None
```
and then doing the if part.